File: Task2/ex2Kotlin/morse.kt
import java.util.*

fun main() {
    // Код Морзе для каждой буквы
    val morseCode = mapOf(
        'a' to ".-", 'b' to "-...", 'c' to "-.-.", 'd' to "-..", 'e' to ".", 'f' to "..-.",
        'g' to "--.", 'h' to "....", 'i' to "..", 'j' to ".---", 'k' to "-.-", 'l' to ".-..",
        'm' to "--", 'n' to "-.", 'o' to "---", 'p' to ".--.", 'q' to "--.-", 'r' to ".-.",
        's' to "...", 't' to "-", 'u' to "..-", 'v' to "...-", 'w' to ".--", 'x' to "-..-",
        'y' to "-.--", 'z' to "--.."
    )

    // Множество для хранения уникальных кодов Морзе
    val uniqueMorseCodes = HashSet<String>()

    // Считываем ввод
    println("Введите слова: ")
    val input = readLine()?.trim() ?: return

    // Разбиваем строку на слова
    val words = input.split(" ").filter { it.isNotEmpty() }

    for (word in words) {
        // Проверяем, что слово состоит только из строчных английских букв
        if (word.any { it !in 'a'..'z' }) {
            println("Ошибка: Введены недопустимые символы (только строчные буквы разрешены).")
            return
        }

        // Преобразуем слово в код Морзе
        val morse = word.map { morseCode[it] ?: "" }.joinToString("")

        // Сортируем полученный код Морзе
        val sortedMorse = morse.toCharArray().sorted().joinToString("")

        // Добавляем отсортированный код в множество
        uniqueMorseCodes.add(sortedMorse)
    }

    // Выводим количество уникальных кодов Морзе
    println("Количество уникальных кодов Морзе: ${uniqueMorseCodes.size}")
}

//java -jar morse.jar 

File: Task2/ex2Rust/my_rust_project/src/morse.rs
use std::collections::HashSet;
use std::io::{self, Write};

fn main() {
    // Карта кодов Морзе для каждой буквы
    let morse_code: std::collections::HashMap<char, &str> = [
        ('a', ".-"), ('b', "-..."), ('c', "-.-."), ('d', "-.."), ('e', "."), ('f', "..-."), ('g', "--."),
        ('h', "...."), ('i', ".."), ('j', ".---"), ('k', "-.-"), ('l', ".-.."), ('m', "--"), ('n', "-."),
        ('o', "---"), ('p', ".--."), ('q', "--.-"), ('r', ".-."), ('s', "..."), ('t', "-"), ('u', "..-"),
        ('v', "...-"), ('w', ".--"), ('x', "-..-"), ('y', "-.--"), ('z', "--..")
    ].iter().cloned().collect();

    // Множество для хранения уникальных кодов Морзе
    let mut unique_morse_codes = HashSet::new();

    // Ввод данных
    print!("Введите слова: ");
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let input = input.trim();

    // Разбиваем строку на слова
    let words: Vec<&str> = input.split_whitespace().collect();

    // Обрабатываем каждое слово
    for word in words {
        // Преобразуем слово в код Морзе
        let morse: String = word.chars().map(|c| {
            *morse_code.get(&c).unwrap_or(&"")  // Получаем код Морзе для каждой буквы
        }).collect();

        // Сортируем символы кода Морзе
        let mut morse_chars: Vec<char> = morse.chars().collect();
        morse_chars.sort();
        let sorted_morse: String = morse_chars.into_iter().collect();

        // Добавляем отсортированный код Морзе в множество
        unique_morse_codes.insert(sorted_morse);
    }

    // Выводим количество уникальных кодов Морзе
    println!("Количество уникальных кодов Морзе: {}", unique_morse_codes.len());
}
